sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1648,16 +1648,56 @@ object RewriteCorrelatedScalarSubquery extends Rule[LogicalPlan] {
   }
 
   /**
+   * Statically evaluate an expression containing one or more aggregates on an empty input.
+   */
+  private def evalOnZeroTups(expr : Expression) : Option[Any] = {
+    // AggregateExpressions are Unevaluable, so we need to replace all aggregates
+    // in the expression with the value they would return for zero input tuples.
+    val rewrittenExpr = expr transform {
+      case a @ AggregateExpression(aggFunc, _, _, resultId) =>
+        val resultLit = aggFunc.defaultResult match {
+          case Some(lit) => lit
+          case None => Literal.default(NullType)
+        }
+        Alias(resultLit, "aggVal") (exprId = resultId)
+    }
+    Option(rewrittenExpr.eval())
+  }
+
+  /**
    * Construct a new child plan by left joining the given subqueries to a base plan.
    */
   private def constructLeftJoins(
       child: LogicalPlan,
       subqueries: ArrayBuffer[ScalarSubquery]): LogicalPlan = {
     subqueries.foldLeft(child) {
       case (currentChild, ScalarSubquery(query, conditions, _)) =>
+        val aggOutputExpr = query.asInstanceOf[Aggregate].aggregateExpressions.head
```

Comment:
I can help you with that. The following construction 'rules' apply:

1. A `ScalarSubquery` has to contain an `Aggregate`. The Aggregate is can be nested in a `Filter` when a `HAVING` clauses is specified. This is the basic `ScalarSubquery`.
2. If the `ScalarSubquery` produces an `Attribute` that is also present in the outer query, then we wrap it with a `Project` and alias the offending attributes in order to prevent attribute clashes between the inner and outer queries.
3. In case of SQL generation we can add further `Project` and `SubqueryAlias` nodes.

The example applies rules 1 & 2.